lib/src/Pos/Binary/Relay.hs

```diff
@@ -4,37 +4,38 @@ module Pos.Binary.Relay () where
 
 import           Universum
 
-import           Pos.Binary.Class                 (Bi (..))
-import           Pos.Binary.Crypto                ()
-import           Pos.Binary.Ssc                   ()
-import           Pos.Binary.Update                ()
-import           Pos.Communication.Types.Relay    (DataMsg (..))
-import           Pos.Core.Configuration           (HasConfiguration)
-import           Pos.Crypto                       (hash)
-import           Pos.Delegation.Types             (ProxySKLightConfirmation)
-import           Pos.Types                        (ProxySKHeavy, ProxySKLight)
-import           Pos.Update.Core                  (UpdateProposal, UpdateVote (..))
+import           Pos.Binary.Class              (Bi (..))
+import           Pos.Binary.Crypto             ()
+import           Pos.Binary.Ssc                ()
+import           Pos.Binary.Update             ()
+import           Pos.Communication.Types.Relay (DataMsg (..))
+import           Pos.Core.Configuration        (HasConfiguration)
+import           Pos.Crypto                    (hash)
+import           Pos.Delegation.Types          (ProxySKLightConfirmation)
+import           Pos.Types                     (ProxySKHeavy, ProxySKLight)
+import           Pos.Update.Core               (UpdateProposal, UpdateVote (..))
 
 instance HasConfiguration => Bi (DataMsg (UpdateProposal, [UpdateVote])) where
-  encode = encode . dmContents
-  decode = do
-    c@(up, votes) <- decode
-    let !id = hash up
-    unless (all ((id ==) . uvProposalId) votes) $ fail "decode@DataMsg@Update: vote's uvProposalId must be equal UpId"
-    pure $ DataMsg c
+    encode = encode . dmContents
+    decode = do
+        c@(up, votes) <- decode
+        let !id = hash up
+        unless (all ((id ==) . uvProposalId) votes) $
```

Comment:
This shouldn't be done when noCheck is on.

Sideline question: if they all have to equal `hash up`, why include this redundant information?

Reply:
Fixed.

Fair. I have no direct answer. Maybe it was just too tedious to optimize: it would require to store modified votes (which, in order, would require us to write `UpdateVote'` that doesn't have `uvProposalId`) plus `hash up`.